### PR TITLE
SCC OpenShift types

### PR DESCRIPTION
**Pull request by pweil-**
Conversion of upstream types into OpenShift types so that we can stop carrying the patch.

@liggitt @pmorie @smarterclayton @deads2k 

I tried to break this up into commits that were easier to review but this is still a large PR.

### CODE REVIEW

**Comment by deads2k** [comment]
This is a **small** one??

**Comment by deads2k** [comment]
My eyes glazed over part way through, but it looks like a straightforward "move".  I was expecting to see a spot where you handled the get/list/update/delete (not create) of sccs from under the kubernetes storage prefix as well as the openshift prefix.  Did I miss it?

**Comment by pweil-** [comment]
Hey, I never said it was small.  The second commit is really just a big move of the types.  The "interesting" commit needs some unglazed eyes though :)

> I was expecting to see a spot where you handled the get/list/update/delete (not create) of sccs from under the kubernetes storage prefix as well as the openshift prefix. Did I miss it?

Not a miss.  Actually, I'm curious about how we want to handle this type of thing.  If we need a migration or some sort of compatibility feature then maybe someone can point me to an existing use case for a pattern to follow.

**Comment by pweil-** [comment]
Also, just as a general FYI, my plan would be to get this in first then do the renaming discussed (move SCC to PodSecurityPolicy) so migration work may want to wait for that so this PR doesn't become even bigger.

**Comment by deads2k** [comment]
> If we need a migration or some sort of compatibility feature then maybe someone can point me to an existing use case for a pattern to follow.

Congrats on being the first.  If we want to support backwards compatibility, then this merging of two different locations is essential.  It will drive you batty though.  You end up having to do unnatural things to prevent conflicting keys as a for instance.

**Comment by liggitt** [comment]
Offhand, the compatability implications are:
* Old clients will still look under api/... for SCC, but they won't exist there any more. @smarterclayton how do we want to handle this?
* During an update, the first master to start with this code will see no SCCs and hit the empty list during the `ensureDefaultSecurityContextConstraints` check. It should probably be the one to migrate from the old etcd key to the new one. That means the data will be duplicated, with new masters seeing it under OpenShift's prefix, and old masters seeing it under K8s' prefix. @smarterclayton do we care about stitching together those storage locations with get, list, watch, etc, or can we say "don't mutate SCC's until you're done with your rolling update"?

**Comment by pweil-** [comment]
@smarterclayton Knowing that the end goal is to try to get this upstream are we creating more problems than solutions with this?  Eventually we'll have to migrate back.

**Comment by pweil-** [comment]
Holding off on rebasing until the above question is answered.

@smarterclayton - thoughts on this issue?

**Comment by smarterclayton** [comment]
On Aug 19, 2015, at 10:48 AM, Jordan Liggitt <notifications@github.com>
wrote:

Offhand, the compatability implications are:

   - Old clients will still look under api/... for SCC, but they won't
   exist there any more. @smarterclayton <https://github.com/smarterclayton>
   how do we want to handle this?


We have to support the old paths.  In this case they are two object kinds
registered under the different Apis.



   - During an update, the first master to start with this code will see no
   SCCs and hit the empty list during the
   ensureDefaultSecurityContextConstraints check. It should probably be the
   one to migrate from the old etcd key to the new one. That means the data
   will be duplicated, with new masters seeing it under OpenShift's prefix,
   and old masters seeing it under K8s' prefix. @smarterclayton
   <https://github.com/smarterclayton> do we care about stitching together
   those storage locations with get, list, watch, etc, or can we say "don't
   mutate SCC's until you're done with your rolling update"?


Use the old storage paths.




—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4252#issuecomment-132604860>.

**Comment by smarterclayton** [comment]
If you think you can deliver the type upstream, do it in parallel.  But I
don't think you're going to get the review bandwidth you need to land it
for 1.1.  You could try for expapi but you'll still need some back compat.

The current rebase solution is untenable I think - it's probably a lesson
we need to take to heart (be very careful what we carry).

On Aug 19, 2015, at 1:24 PM, Paul Weil <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> Knowing that the end
goal is to try to get this upstream are we creating more problems than
solutions with this? Eventually we'll have to migrate back.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4252#issuecomment-132679008>.

**Comment by pweil-** [comment]
@liggitt @smarterclayton could I get some eyes on the last commit here?  I didn't see a nice way to register new storage to an existing service.  You cannot register another REST into the container with the same root.  I added a convenience method to allow me to do it which seems to be working but I want to make sure I didn't miss something obvious.

Testing:
```
[vagrant@openshiftdev origin]$ oc get scc --loglevel=6
I0827 13:34:46.560360   31742 debugging.go:120] GET https://10.0.2.15:8443/oapi/v1/securitycontextconstraints 200 OK in 27 milliseconds
NAME         PRIV      CAPS      HOSTDIR   SELINUX     RUNASUSER
privileged   true      []        true      RunAsAny    RunAsAny
restricted   false     []        false     MustRunAs   MustRunAsRange
[vagrant@openshiftdev origin]$ ~/oldoc get scc --loglevel=6
I0827 13:34:51.918377   31748 debugging.go:120] GET https://10.0.2.15:8443/api 200 OK in 20 milliseconds
I0827 13:34:51.920291   31748 debugging.go:120] GET https://10.0.2.15:8443/api/v1/securitycontextconstraints 200 OK in 1 milliseconds
NAME         PRIV      CAPS      HOSTDIR   SELINUX     RUNASUSER
privileged   true      []        true      RunAsAny    RunAsAny
restricted   false     []        false     MustRunAs   MustRunAsRange

```

**Comment by smarterclayton** [comment]
I think instead of `users` and `groups` we should be using `subjects []ObjectReference` as david did in role bindings.

**Comment by pweil-** [comment]
Still some work left to do but the rename of SecurityContextConstraints to PodSecurityPolicy is complete in this PR.  For backwards compatibility I have maintained an SCC type that embeds a PSP and created a proxy storage implementation that uses the PSP storage for SCCs.  With that I was able to create an SCC with an old client as well as list PSP/SCC with new clients.

https://github.com/pweil-/origin/blob/a0ce07b6055128487065806884aff4040ad1c03c/pkg/security/policy/registry/securitycontextconstraint/proxy.go

Left:

1.  Make the upstream commit for registering to an existing webservice
1.  Comments and tests on the proxy storage
1.  Final sweep on anything still referencing SCC

Follow ups so this doesn't grow any larger:

1.  Change users/groups to subjects

**Comment by pweil-** [comment]
As much as I hate to say this I think we're getting into a risky spot by carrying this as PodSecurityPolicy rather than SCC.  Since the upstream has gained some traction there have been suggestions on renaming some of the fields.  My worry is that if I pull in what I currently have for upstream and we release it then something else changes we'll be carrying conversions from SCC -> PSPv1 -> PSPv2.  

I really don't like the idea of having to go back through this and getting things back to the SCC state but I think it might be the most reasonable route.

@smarterclayton @pmorie @liggitt - opinions on this?

**Comment by pmorie** [comment]
@pweil- I think it's reasonable to wait a little longer to see if we can get some consensus upstream

**Comment by danmcp** [comment]
@pweil- Needs rebase.

**Comment by pweil-** [comment]
This needs some eyeballs.  To avoid them melting, I'm particularly trying to confirm that in the last commit the scc backwards compatibility is correctly implemented.  https://github.com/pweil-/origin/commit/5b8ddb143500b1df3752fc0c1dc2cd4671620e9b#diff-762238b2068471cb41a9d2bf27b176a2R1

**Comment by pweil-** [comment]
This doesn't seem to be getting much action, probably due to size.  I will break this one up into two PRs.  

1.  Straight move of SCC to OpenShift
1.  Intro of the PSP type

**Comment by smarterclayton** [comment]
Can you summarize the latest upstream?  I'll go through and do a review
today if you can help me get context

On Oct 5, 2015, at 9:34 AM, Paul Weil <notifications@github.com> wrote:

This doesn't seem to be getting much action, probably due to size. I will
break this one up into two PRs.

   1. Straight move of SCC to OpenShift
   2. Intro of the PSP type

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4252#issuecomment-145529082>.

**Comment by pweil-** [comment]
Latest upstream is adding PSP to the experimental API (impl details have been removed to try and get the api changes agreed upon).  Here is a summary of the changes

1.  intro of the spec object - not a big deal
1.  more succinct naming - not a big deal
1.  change to support host port ranges - this is a non-compatible change in terms of converting PSP to SCC.  I think in our carry we should maintain the `HostPorts` boolean in PSP until we can drop SCC
1.  intro of `VolumeSecurityPolicy` - this is a compatible change that we can introduce

```go
type PodSecurityPolicySpec struct {
	// Privileged determines if a pod can request to be run as privileged.
	Privileged bool `json:"privileged,omitempty"`
	// Capabilities is a list of capabilities that can be added.
	Capabilities []api.Capability `json:"capabilities,omitempty"`
	// Volumes allows and disallows the use of different types of volume plugins.
	Volumes VolumeSecurityPolicy `json:"volumes,omitempty"`
	// HostNetwork determines if the policy allows the use of HostNetwork in the pod spec.
	HostNetwork bool `json:"hostNetwork,omitempty"`
	// HostPorts determines which host port ranges are allowed to be exposed.
	HostPorts []HostPortRange `json:"hostPorts,omitempty"`
	// HostPID determines if the policy allows the use of HostPID in the pod spec.
	HostPID bool `json:"hostPID,omitempty"`
	// HostIPC determines if the policy allows the use of HostIPC in the pod spec.
	HostIPC bool `json:"hostIPC,omitempty"`
	// SELinuxContext is the strategy that will dictate the allowable labels that may be set.
	SELinuxContext SELinuxContextStrategyOptions `json:"seLinuxContext,omitempty"`
	// RunAsUser is the strategy that will dictate the allowable RunAsUser values that may be set.
	RunAsUser RunAsUserStrategyOptions `json:"runAsUser,omitempty"`

	// The users who have permissions to use this policy
	Users []string `json:"users,omitempty"`
	// The groups that have permission to use this policy
	Groups []string `json:"groups,omitempty"`
}

// HostPortRange defines a range of host ports that will be enabled by a policy
// for pods to use.  It requires both the start and end to be defined.
type HostPortRange struct {
	// Start is the beginning of the port range which will be allowed.
	Start int `json:"start"`
	// End is the end of the port range which will be allowed.
	End int `json:"end"`
}

// VolumeSecurityPolicy allows and disallows the use of different types of volume plugins.
type VolumeSecurityPolicy struct {
	// HostPath allows or disallows the use of the HostPath volume plugin.
	// More info: http://releases.k8s.io/HEAD/docs/user-guide/volumes.md#hostpath
	HostPath bool `json:"hostPath,omitempty"`
	// EmptyDir allows or disallows the use of the EmptyDir volume plugin.
	// More info: http://releases.k8s.io/HEAD/docs/user-guide/volumes.md#emptydir
	EmptyDir bool `json:"emptyDir,omitempty"`
	// GCEPersistentDisk allows or disallows the use of the GCEPersistentDisk volume plugin.
	// More info: http://releases.k8s.io/HEAD/docs/user-guide/volumes.md#gcepersistentdisk
	GCEPersistentDisk bool `json:"gcePersistentDisk,omitempty"`
	// AWSElasticBlockStore allows or disallows the use of the AWSElasticBlockStore volume plugin.
	// More info: http://releases.k8s.io/HEAD/docs/user-guide/volumes.md#awselasticblockstore
	AWSElasticBlockStore bool `json:"awsElasticBlockStore,omitempty"`
	// GitRepo allows or disallows the use of the GitRepo volume plugin.
	GitRepo bool `json:"gitRepo,omitempty"`
	// Secret allows or disallows the use of the Secret volume plugin.
	// More info: http://releases.k8s.io/HEAD/docs/user-guide/volumes.md#secrets
	Secret bool `json:"secret,omitempty"`
	// NFS allows or disallows the use of the NFS volume plugin.
	// More info: http://releases.k8s.io/HEAD/docs/user-guide/volumes.md#nfs
	NFS bool `json:"nfs,omitempty"`
	// ISCSI allows or disallows the use of the ISCSI volume plugin.
	// More info: http://releases.k8s.io/HEAD/examples/iscsi/README.md
	ISCSI bool `json:"iscsi,omitempty"`
	// Glusterfs allows or disallows the use of the Glusterfs volume plugin.
	// More info: http://releases.k8s.io/HEAD/examples/glusterfs/README.md
	Glusterfs bool `json:"glusterfs,omitempty"`
	// PersistentVolumeClaim allows or disallows the use of the PersistentVolumeClaim volume plugin.
	// More info: http://releases.k8s.io/HEAD/docs/user-guide/persistent-volumes.md#persistentvolumeclaims
	PersistentVolumeClaim bool `json:"persistentVolumeClaim,omitempty"`
	// RBD allows or disallows the use of the RBD volume plugin.
	// More info: http://releases.k8s.io/HEAD/examples/rbd/README.md
	RBD bool `json:"rbd,omitempty"`
	// Cinder allows or disallows the use of the Cinder volume plugin.
	// More info: http://releases.k8s.io/HEAD/examples/mysql-cinder-pd/README.md
	Cinder bool `json:"cinder,omitempty"`
	// CephFS allows or disallows the use of the CephFS volume plugin.
	CephFS bool `json:"cephfs,omitempty"`
	// DownwardAPI allows or disallows the use of the DownwardAPI volume plugin.
	DownwardAPI bool `json:"downwardAPI,omitempty"`
	// FC allows or disallows the use of the FC volume plugin.
	FC bool `json:"fc,omitempty"`
}
```

**Comment by pweil-** [comment]
rebased and ready for review.  There are some TODOs in the storage mechanism for item 3 above pending review.  If this looks good then I'd like to finish up testing and ensure that the cluster upgrade would work as expected.  Then follow up with the host pid/ipc integration PR that is waiting review as well.

**Comment by pweil-** [comment]
bump

**Comment by pmorie** [comment]
@pweil- Does it blend?

**Comment by pmorie** [comment]
I cannot believe there's no blender emoji

**Comment by danmcp** [comment]
@pweil- Needs rebase.

**Comment by pweil-** [comment]
> @pweil- Needs rebase.

on hold for post 3.1.  Will do.

**Comment by danmcp** [comment]
@pweil- Bump

**Comment by pweil-** [comment]
closing for now - will revisit after upstream lands
